packages/protocol/src/action/client/ClientSubscriptions.ts

```diff
@@ -0,0 +1,171 @@
+        return this.#subscriptions[Symbol.iterator]();
+    }
+
+    /**
+     * Terminate all subscriptions.
+     */
+    async close() {
+        if (this.#timeout) {
+            this.#timeout.stop();
+            this.#timeout = undefined;
+        }
+        for (const subscription of this.#subscriptions) {
+            subscription.close();
+        }
+    }
+
+    /**
+     * Restart the timeout timer for the current set of active subscriptions.
+     */
+    resetTimer() {
+        const now = Time.nowMs();
+        let nextTimeoutAt: number | undefined;
+
+        // Process each subscription
+        for (const subscription of this.#subscriptions) {
+            // If reading data reports, ignore for timeout purposes
+            if (subscription.isReading) {
+                continue;
+            }
+
+            // Update timeout or expire if timed out
+            let { timeoutAtMs } = subscription;
+            if (timeoutAtMs === undefined) {
+                // Set timeout time
+                timeoutAtMs = subscription.timeoutAtMs = timeoutFor(subscription);
+            } else if (timeoutAtMs < now) {
+                // Timeout
+                this.#timeOut(subscription);
+                continue;
+            }
+
+            // If this is the earliest timeout, record
+            if (nextTimeoutAt === undefined || nextTimeoutAt > timeoutAtMs) {
+                nextTimeoutAt = timeoutAtMs;
+            }
+        }
+
+        // If no subscriptions require timeout, disable timer
+        if (nextTimeoutAt === undefined) {
+            this.#nextTimeoutAt = undefined;
+            this.#timeout?.stop();
+            return;
+        }
+
+        // Create or update timer if not set for correct interval
+        if (nextTimeoutAt !== this.#nextTimeoutAt) {
+            this.#nextTimeoutAt = nextTimeoutAt;
+            if (this.#timeout) {
+                this.#timeout?.stop();
+                this.#timeout.intervalMs = nextTimeoutAt - now;
+            } else {
+                this.#timeout = Time.getTimer("SubscriptionTimeout", nextTimeoutAt - now, this.resetTimer.bind(this));
+            }
+        }
+    }
+
+    #closeOne(subscription: ClientSubscription, cause: CanceledError | TimeoutError) {
+        if (subscription.isClosed) {
+            return;
+        }
+        subscription.isClosed = true;
+        this.#subscriptions.delete(subscription);
+
+        try {
+            subscription.request.closed?.(cause);
+        } catch (e) {
+            logger.error("Error canceling subscription", Diagnostic.strong(subscription.subscriptionId), e);
+        }
+    }
+
+    #timeOut(registration: ClientSubscription) {
+        logger.info(
+            "Subscription",
+            Diagnostic.strong(registration.subscriptionId),
+            "timed out after",
+            Diagnostic.strong(`${timeoutFor(registration)}ms`),
+        );
+
+        this.#closeOne(registration, new TimeoutError());
```

Comment:
This should trigger a reconnection because assumption is that node it offline or something happened ... so .... maybe just add TODO for now because whole
 reconnection management is missing too

Reply:
I added TODO